src/main/java/spoon/reflect/visitor/NameScope.java

```diff
@@ -0,0 +1,219 @@
+		}
+	}
+
+	/**
+	 * Call it for each visited CtElement
+	 * @param parent the parent ConflictFinder
+	 * @param target an element
+	 * @return new {@link NameScope} if `target` element declares new naming scope or null if there is no new scope
+	 */
+	protected static NameScope onElement(NameScope parent, CtElement target) {
+		class Visitor extends CtAbstractVisitor {
+			NameScope finder = null;
+			@Override
+			public void visitCtCompilationUnit(CtCompilationUnit compilationUnit) {
+				//compilation unit items are added in NameScopeOfType, because they depend on the inhertance hierarchy of the type itself
+			}
+			@Override
+			public <T> void visitCtClass(CtClass<T> ctClass) {
+				finder = new NameScopeOfType(parent, ctClass);
+			}
+			@Override
+			public <T> void visitCtInterface(CtInterface<T> intrface) {
+				finder = new NameScopeOfType(parent, intrface);
+			}
+			@Override
+			public <T extends Enum<?>> void visitCtEnum(CtEnum<T> ctEnum) {
+				finder = new NameScopeOfType(parent, ctEnum);
+			}
+			@Override
+			public <A extends Annotation> void visitCtAnnotationType(CtAnnotationType<A> annotationType) {
+				finder = new NameScopeOfType(parent, annotationType);
+			}
+			@Override
+			public <T> void visitCtMethod(CtMethod<T> m) {
+				finder = new SimpleNameScope(parent, m, m.getParameters());
+			}
+			@Override
+			public <T> void visitCtConstructor(CtConstructor<T> c) {
+				finder = new SimpleNameScope(parent, c, c.getParameters());
+			}
+			@Override
+			public <T> void visitCtLambda(CtLambda<T> lambda) {
+				finder = new SimpleNameScope(parent, lambda, lambda.getParameters());
+			}
+			@Override
+			public void visitCtCatch(CtCatch catchBlock) {
+				finder = new SimpleNameScope(parent, catchBlock).addVariable(catchBlock.getParameter());
+			}
+			@Override
+			public <R> void visitCtBlock(CtBlock<R> block) {
+				finder = new SimpleNameScope(parent, block);
+			}
+			@Override
+			public <T> void visitCtLocalVariable(CtLocalVariable<T> localVariable) {
+				if (parent instanceof SimpleNameScope) {
+					((SimpleNameScope) parent).addVariable(localVariable);
+				} else {
+					throw new SpoonException("Cannot add local variable when parent is missing");
+				}
+			}
+		};
+		Visitor scanner = new Visitor();
+		target.accept(scanner);
+		return scanner.finder;
+	}
+
+	private final NameScope parent;
+	private final CtElement scopeElement;
+
+	protected NameScope(NameScope parent, CtElement scopeElement) {
+		this.parent = parent;
+		this.scopeElement = scopeElement;
+	}
+
+	/**
+	 * @return the {@link CtElement} which represents the current scope
+	 */
+	public final CtElement getScopeElement() {
+		return scopeElement;
+	}
+
+	/**
+	 * @param name to be searched simple name
+	 * @param consumer is called for each named element with same name which are accessible from this {@link NameScope}
+	 * 	as long as there are some elements and consumer returns null. If `consumer` return not null value then it is returned
```

Comment:
I suggest fixing the sentence to clarify what is going on. I do not understand what you mean here.